File: backend/app/schemas/social.py
"""Social (feed) request/response schemas (docs/06, docs/10)."""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class Author(BaseModel):
    id: uuid.UUID
    username: str
    display_name: str | None = None
    avatar_url: str | None = None


class PostCreate(BaseModel):
    # Either body or media (or both) must be present — enforced in the route.
    body: str | None = Field(default=None, max_length=5000)
    media: list = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list, max_length=10)
    kind: str = "text"


class PostOut(BaseModel):
    id: uuid.UUID
    author: Author
    kind: str
    body: str | None
    media: list
    tags: list[str]
    like_count: int
    comment_count: int
    liked_by_me: bool
    created_at: datetime
    # Why this post is in the feed ("Followed", "Popular in #legday", …). Read-only.
    reason: str | None = None


class MediaOut(BaseModel):
    url: str
    type: str = "image"


class CommentCreate(BaseModel):
    body: str = Field(min_length=1, max_length=2000)


class CommentOut(BaseModel):
    id: uuid.UUID
    author: Author
    body: str
    created_at: datetime


# ---- Social graph: follows / discover ----


class UserCard(BaseModel):
    """A user as shown in search / suggestions, with the viewer's follow state."""

    id: uuid.UUID
    username: str
    display_name: str | None = None
    avatar_url: str | None = None
    follower_count: int = 0
    is_following: bool = False


class FollowState(BaseModel):
    following: bool
    follower_count: int


class PublicProfile(BaseModel):
    """Another user's profile as seen when 'stalking' from search (read-only)."""

    id: uuid.UUID
    username: str
    display_name: str | None = None
    avatar_url: str | None = None
    age: int | None = None
    sex: str | None = None
    goals: list[str] = []
    follower_count: int = 0
    following_count: int = 0
    post_count: int = 0
    is_following: bool = False
    is_me: bool = False
    # Streaks are None when the user has hidden them (and it isn't you).
    streaks_public: bool = True
    gym_streak: int | None = None
    protein_streak: int | None = None


# ---- Stories ----


class StoryCreate(BaseModel):
    media: list = Field(min_length=1)  # at least one image
    caption: str | None = Field(default=None, max_length=500)


class StoryItem(BaseModel):
    id: uuid.UUID
    author: Author
    media: list
    caption: str | None
    created_at: datetime
    expires_at: datetime
    like_count: int
    comment_count: int
    liked_by_me: bool


class StoryTray(BaseModel):
    """All active stories from one author, grouped for the feed tray."""

    author: Author
    items: list[StoryItem]


class StoryCommentCreate(BaseModel):
    body: str = Field(min_length=1, max_length=2000)


class StoryCommentOut(BaseModel):
    id: uuid.UUID
    author: Author
    body: str
    created_at: datetime

File: frontend/src/api/social.ts
import { api, apiUpload } from "@/api/client";

export interface Author {
  id: string;
  username: string;
  display_name: string | null;
  avatar_url: string | null;
}

export interface Media {
  url: string;
  type: string; // "image"
  w?: number;
  h?: number;
}

export interface Post {
  id: string;
  author: Author;
  kind: string;
  body: string | null;
  media: Media[];
  tags: string[];
  like_count: number;
  comment_count: number;
  liked_by_me: boolean;
  created_at: string;
  reason?: string | null;
}

export interface Comment {
  id: string;
  author: Author;
  body: string;
  created_at: string;
}

export interface UserCard {
  id: string;
  username: string;
  display_name: string | null;
  avatar_url: string | null;
  follower_count: number;
  is_following: boolean;
}

export interface StoryItem {
  id: string;
  author: Author;
  media: Media[];
  caption: string | null;
  created_at: string;
  expires_at: string;
  like_count: number;
  comment_count: number;
  liked_by_me: boolean;
}

export interface StoryTray {
  author: Author;
  items: StoryItem[];
}

export const getFeed = (limit = 10, offset = 0) =>
  api<Post[]>(`/social/feed?limit=${limit}&offset=${offset}`);

export interface NewPost {
  body?: string;
  tags?: string[];
  media?: Media[];
  kind?: string; // "text" (default) | "pr" | ...
}

export const createPost = (post: NewPost) =>
  api<Post>("/social/posts", { method: "POST", body: JSON.stringify(post) });

/** Upload post images (camera/gallery) and get back media descriptors to attach to a post. */
export function uploadPostMedia(files: File[]): Promise<Media[]> {
  const form = new FormData();
  files.forEach((f) => form.append("files", f));
  return apiUpload<Media[]>("/social/media", form);
}

export const toggleLike = (postId: string) =>
  api<{ liked: boolean; like_count: number }>(`/social/posts/${postId}/like`, { method: "POST" });

export const getComments = (postId: string) => api<Comment[]>(`/social/posts/${postId}/comments`);

export const addComment = (postId: string, body: string) =>
  api<Comment>(`/social/posts/${postId}/comments`, {
    method: "POST",
    body: JSON.stringify({ body }),
  });

// ---- Social graph: follows + discover ----

export interface FollowState {
  following: boolean;
  follower_count: number;
}

export const followUser = (userId: string) =>
  api<FollowState>(`/social/users/${userId}/follow`, { method: "POST" });

export const unfollowUser = (userId: string) =>
  api<FollowState>(`/social/users/${userId}/follow`, { method: "DELETE" });

export const searchUsers = (q: string) =>
  api<UserCard[]>(`/social/users/search?q=${encodeURIComponent(q)}`);

export const getSuggestedUsers = () => api<UserCard[]>("/social/users/suggested");

export interface PublicProfile {
  id: string;
  username: string;
  display_name: string | null;
  avatar_url: string | null;
  bio: string | null;
  age: number | null;
  sex: string | null;
  goals: string[];
  follower_count: number;
  following_count: number;
  post_count: number;
  is_following: boolean;
  is_me: boolean;
  streaks_public: boolean;
  gym_streak: number | null;
  protein_streak: number | null;
}

export const getUserProfile = (userId: string) =>
  api<PublicProfile>(`/social/users/${userId}`);

export const getUserPosts = (userId: string, limit = 10, offset = 0) =>
  api<Post[]>(`/social/users/${userId}/posts?limit=${limit}&offset=${offset}`);

// ---- Stories ----

export const getStories = () => api<StoryTray[]>("/social/stories");

export const createStory = (media: Media[], caption?: string) =>
  api<StoryItem>("/social/stories", {
    method: "POST",
    body: JSON.stringify({ media, caption: caption || null }),
  });

export const likeStory = (storyId: string) =>
  api<{ liked: boolean; like_count: number }>(`/social/stories/${storyId}/like`, { method: "POST" });

export const getStoryComments = (storyId: string) =>
  api<Comment[]>(`/social/stories/${storyId}/comments`);

export const addStoryComment = (storyId: string, body: string) =>
  api<Comment>(`/social/stories/${storyId}/comments`, {
    method: "POST",
    body: JSON.stringify({ body }),
  });
